flink-table/flink-table-planner/src/main/java/org/apache/flink/table/factories/TableFactoryUtil.java

```diff
@@ -0,0 +1,137 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.factories;
+
+import org.apache.flink.table.api.TableException;
+import org.apache.flink.table.catalog.ExternalCatalog;
+import org.apache.flink.table.catalog.ExternalCatalogTable;
+import org.apache.flink.table.descriptors.BatchTableDescriptor;
+import org.apache.flink.table.descriptors.Descriptor;
+import org.apache.flink.table.descriptors.StreamTableDescriptor;
+import org.apache.flink.table.sinks.TableSink;
+import org.apache.flink.table.sources.TableSource;
+
+import java.lang.reflect.Method;
+import java.util.Map;
+
+/**
+ * Utility for dealing with {@link TableFactory} using the {@link TableFactoryService}.
+ */
+public class TableFactoryUtil {
+
+	/**
+	 * Returns an external catalog.
+	 */
+	public static ExternalCatalog findAndCreateExternalCatalog(Descriptor descriptor) {
+		Map<String, String> properties = descriptor.toProperties();
+		return TableFactoryService
+			.find(ExternalCatalogFactory.class, properties)
+			.createExternalCatalog(properties);
+	}
+
+	/**
+	 * Returns a table source matching the descriptor.
+	 */
+	public static <T> TableSource<T> findAndCreateTableSource(Descriptor descriptor) {
+		Map<String, String> properties = descriptor.toProperties();
+
+		TableSource<T> tableSource;
+		try {
+			if (descriptor instanceof BatchTableDescriptor || isBatchExternalCatalogTable(descriptor)) {
+
+				Object object = TableFactoryService.find(
+					Class.forName("org.apache.flink.table.factories.BatchTableSourceFactory"),
+					properties);
+				Method method = object.getClass().getDeclaredMethod("createBatchTableSource", Map.class);
+
+				tableSource = (TableSource<T>) method.invoke(object, properties);
+			} else if (descriptor instanceof StreamTableDescriptor || isStreamExternalCatalogTable(descriptor)) {
+
+				Object object = TableFactoryService.find(
+					Class.forName("org.apache.flink.table.factories.StreamTableSourceFactory"),
+					properties);
+				Method method = object.getClass().getDeclaredMethod("createStreamTableSource", Map.class);
+
+				tableSource = (TableSource<T>) method.invoke(object, properties);
+			} else {
+				throw new TableException(
+					String.format(
+						"Unsupported table descriptor: %s",
+						descriptor.getClass().getName())
+				);
+			}
+		} catch (Throwable t) {
+			throw new TableException("findAndCreateTableSource failed.", t);
+		}
+
+		return tableSource;
+	}
+
+	/**
+	 * Returns a table sink matching the descriptor.
+	 */
+	public static <T> TableSink<T> findAndCreateTableSink(Descriptor descriptor) {
+		Map<String, String> properties = descriptor.toProperties();
+
+		TableSink<T> tableSink;
+		try {
+			if (descriptor instanceof BatchTableDescriptor || isBatchExternalCatalogTable(descriptor)) {
+				Object object = TableFactoryService.find(
+					Class.forName("org.apache.flink.table.factories.BatchTableSinkFactory"),
```

Comment:
How about we introduce common interface `TableSink/SourceFactory` for Stream and Batch Factories? This interface would return always `TableSink/Source` not its Stream/Batch counterparts.

This way:

1. We wouldn't need the reflective lookup. We would just have:
```
TableSinkFactory factory = TableFactoryService.find(TableSinkFactory.class, properties);
return factory.createTableSink(...);
```
2. This would already be prepared for unifying Batch and Stream Sinks/Sources in the future.

Reply:
This would be good if we change it to TableSink/Source. However, I'm concern about the compatible problem. These methods are widely used by users. @twalthr What's your opinions here.

Reply:
I am not saying we should change the signature of e.g.`BatchTableSourceFactory` What I am suggesting is the following:

```
interface TableSourceFactory {
   TableSource createTableSource(...)
}

trait BatchTableSourceFactory extends TableSourceFactory {
   def createBatchTableSink(...): BatchTableSink[T]

   override def createTableSource(...) : TableSource = createBatchTableSource(...)
}

trait StreamTableSourceFactory extends TableSourceFactory {
   def createStreamTableSource(...): StreamTableSource[T]

   override def createTableSource(...) : TableSource = createStreamTableSource(...)
}
```

Reply:
Good idea! This is a nice solution.